File: airbyte-integrations/connectors/destination-jdbc/src/test/java/io/airbyte/integrations/destination/jdbc/copy/SwitchingDestinationTest.java
/*
 * Copyright (c) 2022 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination.jdbc.copy;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.common.collect.ImmutableMap;
import io.airbyte.integrations.base.Destination;
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog;
import java.util.Map;
import java.util.function.Consumer;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class SwitchingDestinationTest {

  enum SwitchingEnum {
    INSERT,
    COPY
  }

  private Destination insertDestination;
  private Destination copyDestination;
  private Map<SwitchingEnum, Destination> destinationMap;

  @BeforeEach
  public void setUp() {
    insertDestination = mock(Destination.class);
    copyDestination = mock(Destination.class);
    destinationMap = ImmutableMap.of(
        SwitchingEnum.INSERT, insertDestination,
        SwitchingEnum.COPY, copyDestination);
  }

  @Test
  public void testInsert() throws Exception {
    final var switchingDestination = new SwitchingDestination<>(SwitchingEnum.class, c -> SwitchingEnum.INSERT, destinationMap);

    switchingDestination.getConsumer(mock(JsonNode.class), mock(ConfiguredAirbyteCatalog.class), mock(Consumer.class));

    verify(insertDestination, times(1)).getConsumer(any(), any(), any());
    verify(copyDestination, times(0)).getConsumer(any(), any(), any());

    switchingDestination.check(mock(JsonNode.class));

    verify(insertDestination, times(1)).check(any());
    verify(copyDestination, times(0)).check(any());
  }

  @Test
  public void testCopy() throws Exception {
    final var switchingDestination = new SwitchingDestination<>(SwitchingEnum.class, c -> SwitchingEnum.COPY, destinationMap);

    switchingDestination.getConsumer(mock(JsonNode.class), mock(ConfiguredAirbyteCatalog.class), mock(Consumer.class));

    verify(insertDestination, times(0)).getConsumer(any(), any(), any());
    verify(copyDestination, times(1)).getConsumer(any(), any(), any());

    switchingDestination.check(mock(JsonNode.class));

    verify(insertDestination, times(0)).check(any());
    verify(copyDestination, times(1)).check(any());
  }

}

File: airbyte-cdk/java/airbyte-cdk/db-destinations/src/test/kotlin/io/airbyte/cdk/integrations/destination/jdbc/copy/SwitchingDestinationTest.kt
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */
package io.airbyte.cdk.integrations.destination.jdbc.copy

import com.fasterxml.jackson.databind.JsonNode
import com.google.common.collect.ImmutableMap
import io.airbyte.cdk.integrations.base.Destination
import io.airbyte.protocol.models.v0.ConfiguredAirbyteCatalog
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.mockito.Mockito
import org.mockito.Mockito.mock
import org.mockito.kotlin.any

internal class SwitchingDestinationTest {
    internal enum class SwitchingEnum {
        INSERT,
        COPY
    }

    private lateinit var insertDestination: Destination
    private lateinit var copyDestination: Destination
    private lateinit var destinationMap: Map<SwitchingEnum, Destination>

    @BeforeEach
    fun setUp() {
        insertDestination = Mockito.mock(Destination::class.java)
        copyDestination = Mockito.mock(Destination::class.java)
        destinationMap =
            ImmutableMap.of(
                SwitchingEnum.INSERT,
                insertDestination,
                SwitchingEnum.COPY,
                copyDestination
            )
    }

    @Test
    @Throws(Exception::class)
    fun testInsert() {
        val switchingDestination =
            SwitchingDestination(
                SwitchingEnum::class.java,
                { c: JsonNode -> SwitchingEnum.INSERT },
                destinationMap
            )

        switchingDestination.getConsumer(
            Mockito.mock(JsonNode::class.java),
            Mockito.mock(ConfiguredAirbyteCatalog::class.java),
            mock()
        )

        Mockito.verify(insertDestination, Mockito.times(1)).getConsumer(any(), any(), any())
        Mockito.verify(copyDestination, Mockito.times(0)).getConsumer(any(), any(), any())

        switchingDestination.check(Mockito.mock(JsonNode::class.java))

        Mockito.verify(insertDestination, Mockito.times(1)).check(any())
        Mockito.verify(copyDestination, Mockito.times(0)).check(any())
    }

    @Test
    @Throws(Exception::class)
    fun testCopy() {
        val switchingDestination =
            SwitchingDestination(
                SwitchingEnum::class.java,
                { c: JsonNode -> SwitchingEnum.COPY },
                destinationMap
            )

        switchingDestination.getConsumer(
            Mockito.mock(JsonNode::class.java),
            Mockito.mock(ConfiguredAirbyteCatalog::class.java),
            Mockito.mock()
        )

        Mockito.verify(insertDestination, Mockito.times(0)).getConsumer(any(), any(), any())
        Mockito.verify(copyDestination, Mockito.times(1)).getConsumer(any(), any(), any())

        switchingDestination.check(Mockito.mock(JsonNode::class.java))

        Mockito.verify(insertDestination, Mockito.times(0)).check(any())
        Mockito.verify(copyDestination, Mockito.times(1)).check(any())
    }
}
